src/envvars.zig

```diff
@@ -0,0 +1,311 @@
+///
+/// If the current platform does not have a key specified, all methods that attempt to read the
+/// environment variable will fail at compile time, except for `platformGet` and `platformKey`,
+/// which will return null instead.
+fn platformSpecificNew(
+    comptime T: EnvVarType,
+    comptime posix_key: ?[:0]const u8,
+    comptime windows_key: ?[:0]const u8,
+    comptime opts: EnvVarOpts(T),
+) type {
+    const comptime_key: []const u8 =
+        if (posix_key) |pk| pk else if (windows_key) |wk| wk else "<undefined>";
+
+    if (posix_key == null and windows_key == null) {
+        @compileError("Environment variable " ++ comptime_key ++ " has no keys for POSIX " ++
+            "nor Windows specified. Provide a key for either POSIX or Windows.");
+    }
+
+    const KeyType = [:0]const u8;
+    // Represents the return type of public methods, ignoring the optionality.
+    const NonOptionalReturnType = comptime switch (T) {
+        .string => []const u8,
+        .boolean => bool,
+    };
+
+    // The actual return type of public methods.
+    const ReturnType = if (opts.default != null) NonOptionalReturnType else ?NonOptionalReturnType;
+
+    // Type used to communicate between the Cache and the public interface of this env var.
+    const CachedInterface = union(enum) {
+        /// The environment variable hasn't been loaded yet.
+        undefined: void,
+        /// The environment variable has been loaded but its not set.
+        not_set: void,
+        /// The environment variable is set to a value.
+        value: NonOptionalReturnType,
+    };
+
+    const Cache = switch (T) {
+        .string => struct {
+            const PtrType = ?[*]const u8;
+            const LengthType = u64;
+
+            /// Indicates an environment variable hasn't been loaded yet.
+            const undefined_sentinel: LengthType = std.math.maxInt(LengthType);
+
+            /// Indicates an environment variable isn't set from the outside.
+            var value: std.atomic.Value(PtrType) = std.atomic.Value(PtrType).init(null);
+            var value_len: std.atomic.Value(u64) = .init(std.math.maxInt(LengthType));
+
+            /// Get the cached value.
+            pub inline fn getCached() CachedInterface {
+                const len = value_len.load(.monotonic);
+                if (len == undefined_sentinel) {
+                    return .{ .undefined = {} };
+                }
+
+                const v: PtrType = value.load(.monotonic);
+
+                return if (v) |ptr| .{ .value = ptr[0..len] } else .{ .not_set = {} };
+            }
+
+            pub inline fn reload(raw_env: ?[]const u8) ?NonOptionalReturnType {
+                if (raw_env) |ev| {
+                    value.store(ev.ptr, .monotonic);
+                    value_len.store(ev.len, .monotonic);
+                } else {
+                    value.store(null, .monotonic);
+                    value_len.store(0, .monotonic);
+                }
```

Comment:
There's a race condition here. `getCached` could observe the store to `value_len` in `reload` before the store to `value`, even though the store to `value` happens-before the store to `value_len`. So `getCached` could read the correct len but still load a null pointer, and thus incorrectly return `not_set`.

To fix it:

```suggestion
                const len = value_len.load(.acquire);
                if (len == undefined_sentinel) {
                    return .{ .undefined = {} };
                }

                const v: PtrType = value.load(.monotonic);

                return if (v) |ptr| .{ .value = ptr[0..len] } else .{ .not_set = {} };
            }

            pub inline fn reload(raw_env: ?[]const u8) ?NonOptionalReturnType {
                if (raw_env) |ev| {
                    value.store(ev.ptr, .monotonic);
                    value_len.store(ev.len, .release);
                } else {
                    value.store(null, .monotonic);
                    value_len.store(0, .release);
                }
```